harness/harness-core
# fix: [DEL-4963]: Separate Redis instance for delegate-related operation

## Describe your changes
Summary:
- Separate Redis instance for delegate-related operations and introduce separate Redis Config for delegate service.
- Distributed caching approach using Redisson Client RMap data structure
- Local Redis cache support using RLocalCachedMap
https://harness.atlassian.net/wiki/spaces/DEL/pages/21255423025/Dedicated+Redis+Instance+for+Delegate+Service

Test Performed
Sanity Test with disable flag to use Redis cache
         1) Delegate Installation - CG and NG
         2) K8 end to end flow - CG
         3) NG Pipeline execution

Delegate Cache 
- If flag enabled, Verify cache is present in Redis and value in hash
- Verify Delegate Cache fetch from Redis cache on read and refresh on every 1 minute
- Verify any update on Delegate entity reflects on UI/Redis cache
- Verify delete delegate updates the cache and remove from cache (pending)
- Verify task execution flow with delegate entity fetched from Redis

Delegate Group Cache
- Verify delegate groups cache are getting added to the Redis cache and value in hash
- Verify UI all delegates are displayed under the delegate group
- End to End Task execution with NG pipeline
- Verify list of selectors are displaying correctly in NG   

## Checklist

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
    - CodeFormat: `trigger codeformat`
    - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions


## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)


- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/harness/harness-core/43262)
